feat: public namespace basic function

youabcd opened the pull request:
## What's the purpose of this PR

Created a basic administration page for public namespace

youabcd commented:
I have read the CLA Document and I hereby sign the CLA

nobodyiam commented on apollo-portal/src/main/resources/static/i18n/en.json:
```diff
@@ -316,6 +316,8 @@
   "Index.Topped": "Top Successfully",
   "Index.CancelledFavorite": "Remove favorite successfully",
+  "Index.PublicProject": "Public namespace",
+  "Index.PublicTip": "You haven't created any public projects yet. You can set public projects on the project settings.",
```
I suppose the `public projects` should be `public namespaces`? And public namespaces could be created in the projects page.
```suggestion
  "Index.PublicTip": "You haven't created any public namespaces yet. You can create them in your projects.",
```

nobodyiam commented on apollo-portal/src/main/resources/static/i18n/zh-CN.json:
```diff
@@ -316,6 +316,8 @@
   "Index.GetCreateAppRoleFailed": "获取创建应用权限信息失败",
   "Index.Topped": "置顶成功",
   "Index.CancelledFavorite": "取消收藏成功",
+  "Index.PublicProject": "公共项目",
+  "Index.PublicTip": "您还没有任何公共项目，在项目配置中可以设置哟~",
```
There is still not a proper translation for `Namespace`, so I suggest we keep the English name here. 
```suggestion
  "Index.PublicProject": "公共Namespace",
  "Index.PublicTip": "您还没有任何公共Namespace，在你的项目中可以创建哟~",
```

nobodyiam commented on apollo-portal/src/main/resources/static/scripts/controller/IndexController.js:
```diff
@@ -119,6 +127,23 @@ function IndexController($scope, $window, $translate, toastr, AppUtil, AppServic
             })
     }
 
+    function getPublicProjects() {//TODO
```
I noticed we have a lot of methods/variabled named with `PublicProjects`, however, I think they are actually `PublicNamespaces`? Shall we rename them to better reflect the actual meaning?

nobodyiam reviewed (commented):
Please see some comments below.
BTW, would you please follow the [pull request template](https://github.com/ctripcorp/apollo/blob/master/.github/PULL_REQUEST_TEMPLATE.md) to  add more information about this PR?
It would also be great if some screenshots could be provided.

youabcd commented:
I have made some changes based on suggestions.

nobodyiam commented:
Thanks, I think this is a small milestone achieved!
BTW, I kind of like your original design about the home page by displaying these items in one table, will you keep working on this?

![a](https://user-images.githubusercontent.com/837658/127243201-e651ed11-c965-4220-bb8b-42ef740c2d39.jpg)

youabcd commented:
I have changed the grid display to a table display and also changed the left bar to a menu, as shown below:
![image](https://user-images.githubusercontent.com/63175220/128448434-3d941b02-1b27-4172-9f72-7585ee307778.png)
![image](https://user-images.githubusercontent.com/63175220/128448501-f9de0214-f864-4673-88cd-66a9bc4684b9.png)
![image](https://user-images.githubusercontent.com/63175220/128448527-58f1e2d1-2601-44cb-b2d2-e1fc08775274.png)
![image](https://user-images.githubusercontent.com/63175220/128448562-27e02807-ad30-426f-ab61-206e9da7df07.png)

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,37 +36,221 @@
+                <div class="display-table" ng-show="createdApps.length > 0">
+                    <table>
+                        <tr>
+                            <th style="width: 20%">{{'Index.appTable.appId' | translate }}</th>
```
I think it's better display the information as those in the project page: App Id, App Name, Department, App Owner, Email. And we could reuse the existing labels, e.g. Common.AppId.

![image](https://user-images.githubusercontent.com/837658/129285101-e50b560d-005c-415c-9b2e-147554b07c53.png)

nobodyiam commented on apollo-portal/src/main/resources/static/i18n/en.json:
```diff
@@ -311,11 +311,21 @@
   "Index.CancelledFavorite": "Remove favorite successfully",
+  "Index.PublicNamespace": "Public namespaces",
+  "Index.PublicNamespaceTip": "You haven't created any public namespaces yet. You can create them in your projects.",
+  "Index.appTable.appId": "appId",
```
These labels already exist, e.g, Common.AppId

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,37 +36,221 @@
+                <div class="display-table" ng-show="publicNamespaces.length > 0">
+                    <table>
+                        <tr>
+                            <th style="width: 20%">{{'Index.appTable.appId' | translate }}</th>
```
I think it's better to display these information: Name, AppId, Format, Comment

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,37 +36,221 @@
+                    </table>
+                </div>
+                <div class="no-public text-center" ng-show="!publicNamespaces || publicNamespaces.length == 0">
+                    <h4>{{'Index.FavoriteTip' | translate }}</h4>
```
This looks like wrong?
```suggestion
                    <h4>{{'Index.PublicNamespaceTip' | translate }}</h4>
```

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,37 +36,221 @@
+                        </tr>
+                    </table>
+                </div>
+                <div ng-show="!hasMoreCreatedApps" style="height: 15px"></div>
```
Is this required?

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,37 +36,221 @@
+                <div class="no-favorites text-center" ng-show="!favorites || favorites.length == 0">
+                    <h4>{{'Index.FavoriteTip' | translate }}</h4>
+                </div>
+                <div ng-show="!hasMoreCreatedApps" style="height: 15px"></div>
```
Is this required?

nobodyiam replied:
```suggestion
                <div ng-show="!hasMoreFavorites" style="height: 15px"></div>
```

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,37 +36,221 @@
+                </div>
+            </div>
+            <div class="public-namespace-list" ng-show="whichContent==='3'">
+                <div class="display-table" ng-show="publicNamespaces.length > 0">
```
There might be a lot of public namespaces, I think it's better we provide a search/filter box here.

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,37 +36,221 @@
+                <div class="display-table" ng-show="favorites && favorites.length > 0">
+                    <table>
+                        <tr>
+                            <th style="width: 20%">{{'Index.appTable.appId' | translate }}</th>
```
The same as above, it's better to display the information as those in the project page: App Id, App Name, Department, App Owner, Email.

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,37 +36,221 @@
+                <div class="display-table" ng-show="visitedApps.length > 0">
+                    <table>
+                        <tr>
+                            <th style="width: 20%">{{'Index.appTable.appId' | translate }}</th>
```
The same as above, it's better to display the information as those in the project page: App Id, App Name, Department, App Owner, Email.

nobodyiam reviewed (changes requested):
I tested this feature in my local environment and it looks great! Keep going!

nobodyiam commented:
BTW, I really like the style of your original design, is it possible to adjust the colors, layouts, search boxes to look like below?

![design](https://user-images.githubusercontent.com/837658/127243201-e651ed11-c965-4220-bb8b-42ef740c2d39.jpg)

nobodyiam commented on apollo-portal/src/main/resources/static/i18n/en.json:
```diff
@@ -311,11 +311,17 @@
   "Index.CancelledFavorite": "Remove favorite successfully",
+  "Index.PublicNamespace": "Public namespaces",
+  "Index.PublicNamespaceTip": "You haven't created any public namespaces yet. You can create them in your projects.",
+  "Index.appTable.operation": "operation",
```
```suggestion
  "Index.appTable.operation": "Operation",
```

nobodyiam commented on apollo-portal/src/main/resources/static/styles/demo.css:
```diff
@@ -0,0 +1,539 @@
+/* Logo 字体 */
```
Why is this file needed?

nobodyiam commented on apollo-portal/src/main/resources/static/styles/demo_index.html:
```diff
@@ -0,0 +1,280 @@
+<!DOCTYPE html>
```
Why is this file needed?

nobodyiam commented on apollo-portal/src/main/resources/static/styles/iconfont.css:
```diff
@@ -0,0 +1,31 @@
+@font-face {
```
Please move all iconfront releated to vendor/iconfront folder

nobodyiam commented on apollo-portal/src/main/resources/static/i18n/zh-CN.json:
```diff
@@ -311,11 +311,18 @@
+  "Index.PublicNamespaceTip": "您还没有任何公共Namespace，在你的项目中可以创建哟~",
+  "Index.appTable.operation": "操作",
+  "Index.appTable.Format": "格式",
+  "Index.appTable.Comment": "评论",
```
```suggestion
  "Index.appTable.Comment": "备注",
```

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -28,6 +28,8 @@
     <link rel="stylesheet" type="text/css" media='all' href="vendor/font-awesome.min.css">
 
     <link rel="stylesheet" type="text/css" href="styles/common-style.css">
+    <!--<link rel="stylesheet" type="text/css" href="http://at.alicdn.com/t/font_2742516_pik2lmzsw08.css">-->
```
代码中注释掉的代码统一清理一下吧

nobodyiam commented on apollo-portal/src/main/resources/static/vendor/iconfont/iconfont.js:
```diff
@@ -0,0 +1 @@
+!function(t){var e,n,o,c,i,l='<svg><symbol id="icon-icon" viewBox="0 0 1024 1024"><path d="M898.466935 347.56595 689.623443 347.56595c-39.937594 0-82.428338-30.705326-94.468558-68.273966l-60.770062-189.731202c-12.001335-37.56864-31.857569-37.626969-44.165896-0.155543L427.952792 279.447527c-12.308327 37.453007-54.975079 68.118424-94.912673 68.118424L125.476783 347.56595c-39.937594 0-46.381353 19.2 ...
```
It seems this file is not used?

nobodyiam commented on apollo-portal/src/main/resources/static/vendor/iconfont/iconfont.json:
```diff
@@ -0,0 +1,37 @@
+{
```
It seems this file is not used?

nobodyiam commented on apollo-portal/src/main/resources/static/vendor/iconfont/iconfont.css:
```diff
@@ -0,0 +1,31 @@
+@font-face {
+  font-family: "iconfont"; /* Project id 2742392 */
+  src: url('http://at.alicdn.com/t/font_2742516_pik2lmzsw08.woff2?t=1628826396308') format('woff2'),
```
should we reference the iconfont.woff2, iconfont.woff, iconfont.ttf in the projects instead of external resources?

nobodyiam commented on apollo-portal/src/main/resources/static/scripts/controller/IndexController.js:
```diff
@@ -15,21 +15,142 @@
+        }
+    };
+})
+.directive('inputSearch', function() {
```
I suggest we use the existing select2 component as the search box, just as the top search bar.
You may reference the implementation in directive.js

https://github.com/ctripcorp/apollo/blob/772d2fa862b6622bbeb31b50437680887a35e749/apollo-portal/src/main/resources/static/scripts/directive/directive.js#L37-L87

nobodyiam replied:
@youabcd Are you still working on this? I think we could merge this pr when this issue is resolved

nobodyiam commented on apollo-portal/src/main/resources/static/index.html:
```diff
@@ -36,87 +37,189 @@
-                        <img class="more-img" src="img/more.png" />
-                        <h5>{{'Index.LoadMore' | translate }}</h5>
+                <div ng-show="!hasMoreCreatedApps" style="height: 15px"></div>
+                <div style="width: 180px;margin-left: 30px;margin-top: 20px" ng-show="hasMoreCreatedApps"
```
here is a nice to have suggestion, since we are using table to present the items, it will be nice to add a [pagination](https://angular-ui.github.io/bootstrap/#!#pagination) control here

nobodyiam commented on apollo-portal/src/main/resources/static/scripts/controller/IndexController.js:
```diff
@@ -251,10 +135,25 @@ function IndexController($scope, $window, $translate, toastr, AppUtil, AppServic
         NamespaceService.find_public_namespaces()
             .then(function (result) {
                 $scope.allPublicNamespaces = result;
```
It looks like the current implementation is to load all the data and do the pagination and filter in javascript. It is ok for most of the cases, but if there are a large amount of data, then it might make the browser very slow. What do you think?

nobodyiam replied:
I found the new namespace page also loaded all the public namespaces data and its performance is ok in our [demo site](http://106.54.227.205/namespace.html?#/appid=100004458) so I think we could keep the current implementation.

nobodyiam commented on apollo-portal/src/main/resources/static/scripts/controller/IndexController.js:
```diff
@@ -340,22 +239,14 @@ function IndexController($scope, $window, $translate, toastr, AppUtil, AppServic
-        $scope.$apply();
-        console.log($scope.dataList);
-    };
+        if(data.text.indexOf(params.term) !==-1 ){
```
I think it's better to ignore the case when filtering.

nobodyiam reviewed (changes requested):
Looks much better now. Please find some comments below.

nobodyiam commented on apollo-portal/src/main/resources/static/scripts/controller/IndexController.js:
```diff
@@ -114,11 +126,37 @@ function IndexController($scope, $window, $translate, toastr, AppUtil, AppServic
     }
 
+    function getPublicNamespaces() {
+        var size = 10;
```
this variable seems not used.

nobodyiam commented on apollo-portal/src/main/resources/static/scripts/controller/IndexController.js:
```diff
@@ -114,11 +126,37 @@ function IndexController($scope, $window, $translate, toastr, AppUtil, AppServic
+                });
+                $('#public-name-spaces-search-list').select2({
+                    data: selectResult,
+                    matcher: matchStart,
+                    allowClear:true,
```
I think we don't need to write a custom matcher implementation? Is it ok to use the built-in matcher?
Also I think `allowClear` is not necessary here? Since when user selects some option, we will jump the target page.
```suggestion
```

nobodyiam commented on apollo-portal/src/main/resources/static/scripts/controller/IndexController.js:
```diff
@@ -182,4 +220,33 @@ function IndexController($scope, $window, $translate, toastr, AppUtil, AppServic
+        $scope.whichContent = contentIndex;
+    }
+
+    function matchStart(params, data) {
```
If we don't need a custom matcher, then this function should be removed.

nobodyiam reviewed (changes requested):
Basically I think it's good to go. Please solve the remaining issues and update the latest master branch. Then we could invite more people to review.

BTW, please also update the CHANGES.md.

nobodyiam commented:
@youabcd Looks great now, all the pending issues are resolved! Please update the [CHANGES.md](https://github.com/apolloconfig/apollo/blob/master/CHANGES.md) and I think we are good to go.
